### ISSUE TITLE
A decorated task function that returns generator will fail 

### ISSUE THREAD

**Issue by alfa-th** [PrefectHQ/prefect#8433]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

Returning the result of `pd.read_csv` with `iterator` parameter set to `True` inside a decorated task function will fail the flow. This also raises a `TypeError` from `asyncio`. When i remove the `@task` decorator the bug immediately gone

### Reproduction

```python3
import pandas as pd
from prefect import flow, task
from pathlib import Path


@task(log_prints=True)
def get_generator(path: str):
    df_generator = pd.read_csv(path, iterator=True, chunksize=10000)

    return df_generator


@flow()
def main_flow():
    csv_path = Path() / "yellow_tripdata_2019-02.csv"

    df_gen = get_generator(csv_path)
    df = next(df_gen)
    print(df.head(5))


if __name__ == "__main__":
    main_flow()
```

### Error

```python3
(dtc) user@DESKTOP:~/de-zoomcamp-2023/week2/prefect/prefect_test$ python ingest.py 
18:53:43.697 | INFO    | prefect.engine - Created flow run 'cerulean-lorikeet' for flow 'main-flow'
18:53:43.856 | INFO    | Flow run 'cerulean-lorikeet' - Created task run 'get_generator-0' for task 'get_generator'
18:53:43.857 | INFO    | Flow run 'cerulean-lorikeet' - Executing 'get_generator-0' immediately...
18:53:54.659 | INFO    | Task run 'get_generator-0' - Finished in state Completed()
18:53:54.660 | ERROR   | asyncio - Exception in callback WorkerThread._report_result(<Future pendi...ask_wakeup()]>, None, StopIteration())
handle: <Handle WorkerThread._report_result(<Future pendi...ask_wakeup()]>, None, StopIteration())>
Traceback (most recent call last):
  File "/home/user/anaconda3/envs/dtc/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/user/anaconda3/envs/dtc/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 849, in _report_result
    future.set_exception(exc)
TypeError: StopIteration interacts badly with generators and cannot be raised into a Future
```

### Versions

```Text
Version:             2.7.11
API version:         0.8.4
Python version:      3.10.6
Git commit:          6b27b476
Built:               Thu, Feb 2, 2023 7:22 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         ephemeral
Server:
  Database:          sqlite
  SQLite version:    3.37.2
```

### Additional context

I was trying to write a 700 MB CSV into GCS but i keep getting an OOM error so using pandas generator was one of the solution but the flow keep crashing

**Comment by peytonrunyan** [comment]
I'm pretty sure this is an upstream issue that we can't really do much about at the moment. 

Here's a useful discussion: https://github.com/tiangolo/fastapi/issues/966
Here's the PR where it's fixed in AnyIO: https://github.com/agronholm/anyio/pull/477
Here's the version history, where if you look at the `UNRELEASED` section you'll see `Fixed to_thread.run_sync() hanging on asyncio if the target callable raises StopIteration`: https://github.com/agronholm/anyio/blob/master/docs/versionhistory.rst

I think this should get squared away once this release happens.

**Comment by gerritdeen** [comment]
Hmm, I keep getting a similar error and apparently it should be fixed with anyio 4.0.0, but prefect is still fixed at <4.0.0. Any idea when prefect will start using >4.0.0?

**Comment by serinamarie** [comment]
Hi @gerritdeen, there isn't an issue for that yet that I know of. Would you want to create one?
